src/WebJobs.Script/Binding/HttpBinding.cs

```diff
@@ -45,71 +47,83 @@ public override Collection<CustomAttributeBuilder> GetCustomAttributes(Type para
             }
 
             HttpStatusCode statusCode = HttpStatusCode.OK;
-            JObject headers = null;
+
             if (content is string)
             {
                 try
                 {
-                    // attempt to read the content as a JObject
-                    JObject jo = JObject.Parse((string)content);
-
-                    // if the content is json we capture that so it will be
-                    // serialized as json by WebApi below
-                    content = jo;
-
-                    // TODO: Improve this logic
-                    // Sniff the object to see if it looks like a response object
-                    // by convention
-                    JToken value = null;
-                    if (jo.TryGetValue("body", StringComparison.OrdinalIgnoreCase, out value))
-                    {
-                        content = value;
+                    // attempt to read the content as a ExpandoObject
+                    content = JsonConvert.DeserializeObject<ExpandoObject>((string)content, new ExpandoObjectConverter());
+                }
+                catch (JsonException)
+                {
+                    // not a json response
+                }
+            }
+            
+            IDictionary<string, object> headers = null;
+            ExpandoObject responseObject = content as ExpandoObject;
+            if (responseObject != null)
+            {
+                IDictionary<string, object> cleanResponse = new ExpandoObject();
 
-                        if (value is JValue && ((JValue)value).Type == JTokenType.String)
+                content = cleanResponse;
+                foreach (var pair in responseObject)
+                {
+                    // strip functions
+                    if (!(pair.Value is Delegate))
+                    {
+                        var key = pair.Key;
+                        switch(pair.Key.ToLowerInvariant())
                         {
-                            // convert raw strings so they get serialized properly below
-                            content = (string)value;
-                        }
+                            case "body":
+                                // if there is a body, use it as content instead of cleanResponse
+                                content = pair.Value;
+                                break;
 
-                        if (jo.TryGetValue("headers", StringComparison.OrdinalIgnoreCase, out value) && value is JObject)
-                        {
-                            headers = (JObject)value;
-                        }
+                            case "headers":
+                                headers = pair.Value as IDictionary<string, object>;
```

Comment:
We could process the headers here and remove the call to `ConvertKeysToLowercase` below.